### DISCUSSION TITLE
Do not imply `is:inline` on `<script type="module">`

### DISCUSSION THREAD

**Discussion by hippotastic** [withastro/roadmap#828] [Proposal]
# Summary

This RFC proposes that Astro should treat `<script type="module">` with no other attributes exactly like `<script>`, and not imply `is:inline` anymore. Developers who actually want to use `is:inline` for module scripts can still do so by adding the attribute explicitly.

# Background & Motivation

We have received multiple support questions from developers who are wondering why their `import` statements in client-side scripts inside `.astro` files cannot be resolved. Examining the questions shows that they are pasting example JS code wrapped in a `<script type="module">` element.

This does not work because adding any attributes to a `<script>` element currently causes Astro to also imply `is:inline`, which skips the bundling that would be required for TypeScript and package imports to work.

The current behavior is unexpected if `type="module"` is the only attribute specified. As Astro already treats `<script>` tags without any attributes as if `type="module"` were specified, it should not make any difference to explicitly set this attribute.

# Goals

- Increase the chance that JS module code pasted from examples on the web "just works" in `.astro` files.
- Remove a common source of confusion for developers, especially those new to Astro.
- Keep an escape hatch for developers who actually want to use `is:inline` on script modules by adding the attribute explicitly (`<script type="module" is:inline>`).

# Examples

This is one example of what users are attempting to do:

```html
<script type="module">
  import { gsap } from "gsap";
  import { TextPlugin } from "gsap/TextPlugin";
  gsap.registerPlugin(TextPlugin);
  gsap.defaults({ease: "none"});
  // ...
</script>
```

This currently fails with the following error:

> Uncaught TypeError: Failed to resolve module specifier "gsap". Relative references must start with either "/", "./", or "../".

**Comment by lilnasy** [comment]
Love it! The big question is could we make this a non-breaking change? It is too soon for astro 5.0.

**Reply by Fryuni** [reply]
From what we discussed about it on Discord, I don't think we can. A minimal breaking example would be:

```html
<script type="module">
import something from '//path/relative/to/public/not/to/the/current/astro/file';
</script>
```

That works today, and I don't see how that would not break with this change. It is a simple fix, just add the `is:inline` there, but it is breaking nonetheless.

**Reply by hippotastic** [reply]
Do we have any usage metrics on how common this scenario currently is? And in case it's barely being used, could this justify making it a minor update anyway?

**Reply by martrapp** [reply]
Just my 2ct.

When I first heard about Hippo's proposal, I was excited because I thought it would simplify the convoluted rules around auto-inline and hoisted scripts.

When I think about it, it probably makes things worse:

Imho the main cause of the confusion is auto-hoisting and its consequences. Don't get me wrong, I love writing TypeScript directly in a simple `<script>` tag. But the consequence is that `<script>` looks like HTML, but behaves completely differently. Looking at the history, I can understand how things got to be what they are, but it's hard to explain the semantics of Astro scripts to newbies.

Currently we have
|#|you type|you get|
|-|-------|-------|
1|`<script>...</script>`|`<script type="module" src="..."/>`
2|`<script type="module">...</script>`|`<script type="module">...</script>`
3|`<script type="module" is:inline>...</script>`| `<script type="module">...</script>`
4|`<body><script src="..."/>`|`<head><script type="module" src="..."/>`

Now look at line two. 
This is a) one way to define a deferred inline script in HTML and Astro; and b) the only example here where you get what you type.

The proposal discussed here changes both, a) and b): you don't get a delayed inline script and you don't get what you type, and you have to use the incantation from line 3 to get the earlier effect of line 2.

|#|you type|you get|
|-|-------|-------|
1|...|...
2|`<script type="module">...</script>`|`<script type="module" src="..."/>`
3|`<script type="module" is:inline>...</script>`| `<script type="module">...</script>`
4|...|...|

`<sarcasm>` We can now either be glad that we got rid of the last case where you got what you wrote `</sarcasm>`, or we could turn the discussion to making Astro `<scripts>` more compatible with what people know from HTML. Maybe the now deprecated but still working `hoist` attribute wasn't the worst approach to hoist scripts. 

Or we let things as they are as it is already hard to explain. After all, people just have to drop `type="module"` in the original problem description?

**Reply by hippotastic** [reply]
To me, this is more a decision on whether the benefits outweigh the potential drawbacks. We currently need to guide users through dropping `type="module"` from their code in support issues on a regular basis, so I see a clear benefit.

On the other hand, is it really a drawback that this proposal introduces a new case where there is a difference between user input and rendered HTML output? From my point of view, that Astro processes your code is not unexpected behavior at all. Astro already performs countless processing steps behind the scenes that benefit users right now, with the goal of making simple user input and code pasted from the web "just work" and be performant. This would just be in line with this goal.
